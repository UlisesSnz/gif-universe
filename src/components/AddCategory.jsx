import { useState } from 'react';

export const AddCategory = ({ onAddNewCategory }) => {

    const [ valueInput, setValueInput ] = useState('');
    const [ isValid, setIsValid ] = useState(true);

    const onNewCategory = (e) => {
        setValueInput(e.target.value);
    }

    const onAddCategory = (e) => {
        e.preventDefault();
        const newCategory = valueInput.trim().toLowerCase();
        newCategory === '' ? setIsValid(false) : (
            onAddNewCategory(newCategory),
            setIsValid(true),
            e.target.querySelector('input').blur()
        );
        setValueInput('');
    }

    return (
        <form onSubmit={ onAddCategory } className='mb-md-5 mb-3'>
            <div className="form-floating mb-3">
                <input
                    type="text"
                    className={ isValid ? 'form-control' : 'form-control is-invalid' }
                    id="floatingInputValue"
                    placeholder="Search gifs"
                    value={ valueInput }
                    onChange={ onNewCategory }
                />
                <label htmlFor="floatingInputValue">Search gifs</label>
            </div>
        </form>
    )
}