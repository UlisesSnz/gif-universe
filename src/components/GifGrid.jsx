import PropTypes from 'prop-types';
import { NavDropdown, Row } from 'react-bootstrap';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifItem, LoadingMessage } from './index';

export const GifGrid = ({ category, handleDeleteCategories, handleDeleteCategory }) => {

    const { dataGifs, isLoading } = useFetchGifs(category);

    return (
        <>
            {
                isLoading ? <LoadingMessage category={ category } /> :
                <>
                    {/* Title category and options*/}
                    <div className="h3 pb-2 mb-4 border-3 border-bottom border-dark-subtle" data-id={ category } data-testid={ category }>
                        <NavDropdown
                            title={ category.charAt(0).toUpperCase() + category.slice(1) }
                            menuVariant="dark">
                            <NavDropdown.Item onClick={ handleDeleteCategory } data-testid="btnDeleteCategory">{ `Delete ${category} category` }</NavDropdown.Item>
                            <NavDropdown.Divider />
                            <NavDropdown.Item onClick={ handleDeleteCategories } data-testid="btnDeleteCategories">Delete all categories</NavDropdown.Item>
                        </NavDropdown>
                    </div>

                    {/* Show all gifs of the category */}
                    <Row xs={1} md={2} lg={3} className="g-4 mb-4 mb-md-5">
                    {
                        dataGifs.length === 0 ? <h3 className='text-body-secondary'>{ `Not results for "${category}" :(` }</h3> :
                        dataGifs.map(dataGif => (
                            <GifItem
                                key={ dataGif.id }
                                { ...dataGif }
                            />
                        ))
                    }
                    </Row>
                </>
            }
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired,
    handleDeleteCategories: PropTypes.func.isRequired,
    handleDeleteCategory: PropTypes.func.isRequired,
}