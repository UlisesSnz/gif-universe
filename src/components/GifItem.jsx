import { useState } from 'react';
import { Button, Card, Col } from 'react-bootstrap';
import { GifModal, ToastNotification } from './index';
import { handleDeleteGif, handleDownloadGif } from '../helpers/index';

export const GifItem = ({ id, title, url }) => {

    const [ showModal, setShowModal ] = useState(false);
    const [ showToast, setShowToast ] = useState(false);
    const [ resDownload, setResDownload ] = useState('');

    const handleCloseModal = () => setShowModal(false);
    const handleShowModal = () => setShowModal(true);

    const handleCloseToast  = () => setShowToast(false);
    const handleShowToast = async () => {
        setResDownload(await handleDownloadGif(title, url));
        setShowToast(true);
    }

    return (
        <Col data-id={ id }>
            <Card className="mb-3" style={{ height: '26rem' }} >
                <Card.Img variant="top" src={ url } style={{ height: '18rem' }} alt={ title } />
                <Card.Body className="card-body bg-dark bg-gradient text-white">
                    <Card.Title >{ title }</Card.Title>
                    <Button onClick={ handleShowModal } className="btn-light btn-sm me-2">Expand</Button>
                    <Button onClick={ handleDeleteGif } className="btn-danger btn-sm">Delete</Button>
                    <GifModal
                        title={ title }
                        url={ url }
                        showModal={ showModal }
                        handleCloseModal={ handleCloseModal }
                        handleShowToast={ handleShowToast }
                    />
                    <ToastNotification
                        title={ title }
                        type={ resDownload }
                        showToast={ showToast }
                        handleCloseToast={ handleCloseToast }
                    />
                </Card.Body>
            </Card>
        </Col>
    )
}