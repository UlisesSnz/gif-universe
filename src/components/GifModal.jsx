import PropTypes from 'prop-types';
import { Button, Image, Modal } from 'react-bootstrap';

export const GifModal = ({ title, url, showModal, handleCloseModal, handleShowToast }) => {

    return (
        <Modal show={ showModal } onHide={ handleCloseModal } size="lg" backdrop="static" aria-labelledby="contained-modal-title-vcenter" centered>
            <Modal.Header closeButton className="text-white bg-dark bg-gradient">
                <Modal.Title id="contained-modal-title-vcenter">{ title }</Modal.Title>
            </Modal.Header>
            <Modal.Body>
                <Image width={'100%'} src={ url } />
            </Modal.Body>
            <Modal.Footer>
                <Button variant="secondary" onClick={ handleCloseModal }>
                    Close
                </Button>
                <Button variant="success" onClick={ handleShowToast }>
                    Download
                </Button>
            </Modal.Footer>
        </Modal>
    )
}

GifModal.propTypes = {
    title: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
    showModal: PropTypes.bool.isRequired,
    handleCloseModal: PropTypes.func.isRequired,
    handleShowToast: PropTypes.func.isRequired,
}