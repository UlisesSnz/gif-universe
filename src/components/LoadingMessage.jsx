import PropTypes from 'prop-types';

export const LoadingMessage = ({ category }) => {
  return (
    <>
        <h2 className='text-body-secondary text-center'>{ `Exploring "${category}" in the Gif-Verse` }</h2>
        <div className="sk-cube-grid">
            <div className="sk-cube sk-cube1"></div>
            <div className="sk-cube sk-cube2"></div>
            <div className="sk-cube sk-cube3"></div>
            <div className="sk-cube sk-cube4"></div>
            <div className="sk-cube sk-cube5"></div>
            <div className="sk-cube sk-cube6"></div>
            <div className="sk-cube sk-cube7"></div>
            <div className="sk-cube sk-cube8"></div>
            <div className="sk-cube sk-cube9"></div>
        </div>
    </>
  )
}

LoadingMessage.propTypes = {
  category: PropTypes.string.isRequired,
}