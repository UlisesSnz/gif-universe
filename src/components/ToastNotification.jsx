import PropTypes from 'prop-types';
import { Col, Row, Toast } from 'react-bootstrap';
import alertImage from '../images/alert.png';
import succesImage from '../images/success.png';

export const ToastNotification = ({ title, type, showToast, handleCloseToast }) => {

    return (
        <Row>
            <Col xs={6}>
                <Toast onClose={handleCloseToast} show={showToast} bg={type} delay={5000} autohide className="position-fixed bottom-0 end-0 m-3">
                    <Toast.Header>
                        <img
                            src={ type === 'success' ? succesImage.string : alertImage.string }
                            className="rounded me-2"
                            alt=""
                            style={{ maxWidth: '5%', height: 'auto' }}
                        />
                        <strong className="me-auto" aria-label='strong'>{ type }</strong>
                        <small>Now</small>
                    </Toast.Header>
                    <Toast.Body>
                        { type === 'success' ? `${title} has been downloaded` : `${title} could not be downloaded` }
                    </Toast.Body>
                </Toast>
            </Col>
        </Row>
    )
}

ToastNotification.propTypes = {
    title: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    showToast: PropTypes.bool.isRequired,
    handleCloseToast: PropTypes.func.isRequired,
}