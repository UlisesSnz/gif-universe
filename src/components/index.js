export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';
export * from './GifModal';
export * from './LoadingMessage';
export * from './ToastNotification';