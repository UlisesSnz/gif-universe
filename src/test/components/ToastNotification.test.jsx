import { render, screen } from '@testing-library/react';
import { ToastNotification } from '../../components/ToastNotification';

describe('Pruebas en <ToastNotification />', () => {
  
    const title = 'developer';
    const type = 'success';

    test('debe hacer match con el snapshot', () => {

        const { container } = render( <ToastNotification title={ title } type={ type } showToast={ true } handleCloseToast={ () => {} } /> );

        expect( container ).toMatchSnapshot();

    });

    test('debe de mostrar el tipo de notificaciÃ³n en el componente', () => {

        render( <ToastNotification title={ title } type={ type } showToast={ true } handleCloseToast={ () => {} } /> );

        expect( screen.getByText( type ) ).toBeTruthy();

    });
    
    test('debe de mostrar el mensaje de descarga exitosa', () => {

        render( <ToastNotification title={ title } type={ type } showToast={ true } handleCloseToast={ () => {} } /> );

        expect( screen.getByText( `${title} has been downloaded` ) ).toBeTruthy();

    });

    test('debe de mostrar el mensaje de descarga fallida', () => {

        render( <ToastNotification title={ title } type={ 'danger' } showToast={ true } handleCloseToast={ () => {} } /> );

        expect( screen.getByText( `${title} could not be downloaded` ) ).toBeTruthy();

    });

    test('no debe de mostrar el toast', () => {

        render( <ToastNotification title={ title } type={ type } showToast={ false } handleCloseToast={ () => {} } /> );

        expect( screen.queryAllByText( title ).length ).toBe(0);
        expect( screen.queryAllByText( type ).length ).toBe(0);

    });

})
