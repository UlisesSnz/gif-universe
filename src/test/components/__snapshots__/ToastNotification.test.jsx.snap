// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <ToastNotification /> debe hacer match con el snapshot 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-6"
    >
      <div
        aria-atomic="true"
        aria-live="assertive"
        class="fade toast position-fixed bottom-0 end-0 m-3 bg-success show"
        role="alert"
      >
        <div
          class="toast-header"
        >
          <img
            alt=""
            class="rounded me-2"
            src="string"
            style="max-width: 5%; height: auto;"
          />
          <strong
            aria-label="strong"
            class="me-auto"
          >
            success
          </strong>
          <small>
            Now
          </small>
          <button
            aria-label="Close"
            class="btn-close"
            data-dismiss="toast"
            type="button"
          />
        </div>
        <div
          class="toast-body"
        >
          developer has been downloaded
        </div>
      </div>
    </div>
  </div>
</div>
`;
